vj4/handler/record.py

```diff
@@ -86,3 +91,36 @@ class RecordDetailView(base.Handler):
     rdoc['udoc'], rdoc['pdoc'] = await asyncio.gather(
       user.get_by_uid(rdoc['uid']), problem.get(rdoc['domain_id'], rdoc['pid']))
     self.render('record_detail.html', rdoc=rdoc)
+
+
+@app.route(r'/records/{rid}/pretest_data', 'record_pretest_data')
+class RecordPretestDataView(base.Handler):
+  @base.require_priv(builtin.PRIV_READ_PRETEST_DATA)
+  @base.route_argument
+  @base.sanitize
+  async def get(self, *, rid: objectid.ObjectId):
+    rdoc = await record.get(rid)
+    if not rdoc:
+      raise error.RecordNotFoundError(rid)
+    ddoc = await document.get(rdoc['domain_id'], document.TYPE_PRETEST_DATA, rdoc['data_id'])
+    if not ddoc:
+      raise error.ProblemDataNotFoundError(rdoc['pid'])
+
+    output_buffer = io.BytesIO()
+    zip_file = zipfile.ZipFile(output_buffer, 'a', zipfile.ZIP_DEFLATED)
+    config_content = str(len(ddoc['data_input'])) + "\n"
+    for i, (data_input, data_output) in enumerate(zip(ddoc['data_input'], ddoc['data_output'])):
+      input_file = 'input' + str(i) + '.txt'
+      output_file = 'output' + str(i) + '.txt'
+      config_content += input_file + '|' + output_file + '|' + str("1|10|1024\n")
+      zip_file.writestr('Input/' + input_file, data_input)
+      zip_file.writestr('Output/' + output_file, data_output)
+    zip_file.writestr('Config.ini', config_content)
+
+    # mark all files as created in Windows
```

Comment:
it just marks the `create_system` property ("version made by") of the zip entry. it's up to zip utilities to determine what to do with different values of this property. as far as I know, this property may affect permissions and line endings.